### PR TITLE
Adv/katsura ideals

### PR DESCRIPTION
**Pull request by antonydellavecchia**
Function for generating katsura ideals given a parameter n

```
julia> katsura(3)
ideal(x1 + 2*x2 + 2*x3 + 2*x4 - 1, x1^2 - x1 + 2*x2^2 + 2*x3^2 + 2*x4^2, 2*x1*x2 + 2*x2*x3 - x2 + 2*x3*x4, 2*x1*x3 + x2^2 + 2*x2*x4 - x3)

```

### CODE REVIEW

File: src/Rings/special_ideals.jl [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    return e
+end
+
+@doc Markdown.doc"""
+    katsura(n::Int64)
+
+Given a natural number `n` returns the Katsura ideal generated by\
+$u_m - \sum_{l=n}^n u_{l-m} u_l $, $ 1 - \sum_{l = -n}^n u_l $
+where $u_{-i} = u_i$, and $u_i = 0$ for $i > n$ and $m \in \{-n, \ldots, n\}$. \
```
Comment (thofma) [inline src/Rings/special_ideals.jl:20]: Why the `\` at the end of the line?

**Review by thofma — COMMENTED** [review]
Cool stuff. For these kind of functions we usually also provide a method of the form
```
katsura(R::MPolyRing, n::Int)
```
Since the ideal has generators over ZZ, we can even allow arbitrary coefficient rings.

File: src/Rings/special_ideals.jl (hunk 1 above)
Comment (antonydellavecchia) [reply src/Rings/special_ideals.jl:20]: to break the line so the doc string looks a bit nicer

File: src/Rings/special_ideals.jl [hunk 2]
```diff
@@ -0,0 +1,57 @@
+export katsura
+
+function _get_katsura_exponent(n::Int64, m::Int64, l::Int64)
+    e = zeros(Int64, n + 1)
```
Comment (thofma) [inline src/Rings/special_ideals.jl]:
```suggestion
    e = zeros(Int, n + 1)
```

File: src/Rings/special_ideals.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    else
+        e[abs(l) + 1] = 1
+        e[abs(m - l) + 1] = 1
+    end
+
+    return e
+end
+
+@doc Markdown.doc"""
+    katsura(n::Int64)
```
Comment (thofma) [inline src/Rings/special_ideals.jl]:
```suggestion
    katsura(n::Int)
```

File: src/Rings/special_ideals.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+where $u_{-i} = u_i$, and $u_i = 0$ for $i > n$ and $m \in \{-n, \ldots, n\}$. \
+Also note that indices have been shifted to start from 1.
+
+# Example
+```jldoctest
+julia> katsura(3)
+ideal(x1 + 2*x2 + 2*x3 + 2*x4 - 1, x1^2 - x1 + 2*x2^2 + 2*x3^2 + 2*x4^2, 2*x1*x2 + 2*x2*x3 - x2 + 2*x3*x4, 2*x1*x3 + x2^2 + 2*x2*x4 - x3)
+```
+"""
+function katsura(n::Int64)
```
Comment (thofma) [inline src/Rings/special_ideals.jl]:
```suggestion
function katsura(n::Int)
```

File: src/Rings/special_ideals.jl [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    R, x = PolynomialRing(ZZ, n + 1)
+    polys = Vector{fmpz_mpoly}()
+    coeffs_vec = 2 * ones(fmpz, n + 1)
+    coeffs_vec[1] = 1
+    mono_exps = Matrix{Int64}(identity_matrix(ZZ, n + 1))
```
Comment (thofma) [inline src/Rings/special_ideals.jl]:
```suggestion
    mono_exps = Matrix{Int}(identity_matrix(ZZ, n + 1))
```

File: src/Rings/special_ideals.jl [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        polynomial = MPolyBuildCtx(R)
+        for l in -n:n
+            if abs(l) > n || abs(m - l) > n 
+                continue
+            end
+            e = _get_katsura_exponent(n, m, l)
+            push_term!(polynomial, ZZ(1), e)
+        end
+
+        e = zeros(Int64, n + 1)
```
Comment (thofma) [inline src/Rings/special_ideals.jl]:
```suggestion
        e = zeros(Int, n + 1)
```

File: src/Rings/special_ideals.jl (hunk 1 above)
Comment (thofma) [reply src/Rings/special_ideals.jl:20]: OK, I see. You mean when displayed in the REPL or the HTML docs?

**Comment by thofma** [comment]
Ah, sorry for the misunderstanding. I mean accepting the polynomial ring. This is kind of important so that the ideal can live over a given polynomial ring (otherwise a fresh one will be constructed which is not necessarily what one wants).

**Comment by antonydellavecchia** [comment]
> Ah, sorry for the misunderstanding. I mean accepting the polynomial ring. This is kind of important so that the ideal can live over a given polynomial ring (otherwise a fresh one will be constructed which is not necessarily what one wants).

ah, so it should be an optional parameter, and in this case the n would be implicit in the parent, should I also allow arbitrary coeffs?

**Comment by thofma** [comment]
To have it consistent with other parts, e.g. https://github.com/oscar-system/Oscar.jl/blob/master/src/Combinatorics/SimplicialComplexes.jl#L272-L287, I suggest to have the two methods
```
karatsura(R::MPolyRing)
```
and
```
karatsura(n::Int)
```
The second should just call `karatsura(PolynomialRing(ZZ, n...), ...)`.

**Review by ederc — APPROVED** [review]

**Comment by ederc** [comment]
Note that this is also the correct definition of `katsura(n)`(in contrast to the one used, for example, in Singular): `n`does not denote the number of variables but the number of solutions which is `2^n` for `katsura(n)`.

File: src/Rings/special_ideals.jl [hunk 7]
```diff
@@ -26,32 +26,49 @@ julia> katsura(3)
...
-    coeffs_vec[1] = R(1)
+function katsura(n::Int)
+    R, _ = PolynomialRing(ZZ, n + 1)
+    return katsura(R)
+end
+
+@doc Markdown.doc"""
+    katsura(R::MPolyRing)
+
+Returns the Katsura in the given Polynomial Ring R.
```
Comment (thofma) [inline src/Rings/special_ideals.jl]:
```suggestion
Returns the Katsura ideal in the given Polynomial Ring R.
```

**Comment by thofma** [comment]
Thanks. I will merge once green.

File: src/Rings/special_ideals.jl (hunk 1 above)
Comment (fingolfin) [reply src/Rings/special_ideals.jl:20]: Actually I still don't understand at all why this is beneficial? I've just tried locally the variant without those `\` and find the resulting superior in both terminal and HTML...

File: src/Rings/special_ideals.jl [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+@doc Markdown.doc"""
+    katsura(R::MPolyRing)
+
+Returns the Katsura ideal in the given Polynomial Ring R.
+
+julia> R, _ = QQ["x", "y", "z"]
+(Multivariate Polynomial Ring in x, y, z over Rational Field, fmpq_mpoly[x, y, z])
+
+julia> katsura(R)
+ideal(x + 2*y + 2*z - 1, x^2 - x + 2*y^2 + 2*z^2, 2*x*y + 2*y*z - y)
```
Comment (fingolfin) [inline src/Rings/special_ideals.jl:43]: This is missing code fences around the example code...

I'll submit a PR trying to clean this up

**Pull request merged**